File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/CommodityPhysicalAmericanExercise.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CommodityPhysicalAmericanExercise : Exercise
    {
        public CommodityPhysicalAmericanExercise(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode commencementDatesNode = xmlNode.SelectSingleNode("commencementDates");
            
            if (commencementDatesNode != null)
            {
                if (commencementDatesNode.Attributes["href"] != null || commencementDatesNode.Attributes["id"] != null) 
                {
                    if (commencementDatesNode.Attributes["id"] != null) 
                    {
                        commencementDatesIDRef_ = commencementDatesNode.Attributes["id"].Value;
                        AdjustableOrRelativeDates ob = new AdjustableOrRelativeDates(commencementDatesNode);
                        IDManager.SetID(commencementDatesIDRef_, ob);
                    }
                    else if (commencementDatesNode.Attributes["href"] != null)
                    {
                        commencementDatesIDRef_ = commencementDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        commencementDates_ = new AdjustableOrRelativeDates(commencementDatesNode);
                    }
                }
                else
                {
                    commencementDates_ = new AdjustableOrRelativeDates(commencementDatesNode);
                }
            }
            
        
            XmlNode expirationDatesNode = xmlNode.SelectSingleNode("expirationDates");
            
            if (expirationDatesNode != null)
            {
                if (expirationDatesNode.Attributes["href"] != null || expirationDatesNode.Attributes["id"] != null) 
                {
                    if (expirationDatesNode.Attributes["id"] != null) 
                    {
                        expirationDatesIDRef_ = expirationDatesNode.Attributes["id"].Value;
                        AdjustableOrRelativeDates ob = new AdjustableOrRelativeDates(expirationDatesNode);
                        IDManager.SetID(expirationDatesIDRef_, ob);
                    }
                    else if (expirationDatesNode.Attributes["href"] != null)
                    {
                        expirationDatesIDRef_ = expirationDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expirationDates_ = new AdjustableOrRelativeDates(expirationDatesNode);
                    }
                }
                else
                {
                    expirationDates_ = new AdjustableOrRelativeDates(expirationDatesNode);
                }
            }
            
        
            XmlNode relativeCommencementDatesNode = xmlNode.SelectSingleNode("relativeCommencementDates");
            
            if (relativeCommencementDatesNode != null)
            {
                if (relativeCommencementDatesNode.Attributes["href"] != null || relativeCommencementDatesNode.Attributes["id"] != null) 
                {
                    if (relativeCommencementDatesNode.Attributes["id"] != null) 
                    {
                        relativeCommencementDatesIDRef_ = relativeCommencementDatesNode.Attributes["id"].Value;
                        CommodityRelativeExpirationDates ob = new CommodityRelativeExpirationDates(relativeCommencementDatesNode);
                        IDManager.SetID(relativeCommencementDatesIDRef_, ob);
                    }
                    else if (relativeCommencementDatesNode.Attributes["href"] != null)
                    {
                        relativeCommencementDatesIDRef_ = relativeCommencementDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        relativeCommencementDates_ = new CommodityRelativeExpirationDates(relativeCommencementDatesNode);
                    }
                }
                else
                {
                    relativeCommencementDates_ = new CommodityRelativeExpirationDates(relativeCommencementDatesNode);
                }
            }
            
        
            XmlNode relativeExpirationDatesNode = xmlNode.SelectSingleNode("relativeExpirationDates");
            
            if (relativeExpirationDatesNode != null)
            {
                if (relativeExpirationDatesNode.Attributes["href"] != null || relativeExpirationDatesNode.Attributes["id"] != null) 
                {
                    if (relativeExpirationDatesNode.Attributes["id"] != null) 
                    {
                        relativeExpirationDatesIDRef_ = relativeExpirationDatesNode.Attributes["id"].Value;
                        CommodityRelativeExpirationDates ob = new CommodityRelativeExpirationDates(relativeExpirationDatesNode);
                        IDManager.SetID(relativeExpirationDatesIDRef_, ob);
                    }
                    else if (relativeExpirationDatesNode.Attributes["href"] != null)
                    {
                        relativeExpirationDatesIDRef_ = relativeExpirationDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        relativeExpirationDates_ = new CommodityRelativeExpirationDates(relativeExpirationDatesNode);
                    }
                }
                else
                {
                    relativeExpirationDates_ = new CommodityRelativeExpirationDates(relativeExpirationDatesNode);
                }
            }
            
        
            XmlNode latestExerciseTimeNode = xmlNode.SelectSingleNode("latestExerciseTime");
            
            if (latestExerciseTimeNode != null)
            {
                if (latestExerciseTimeNode.Attributes["href"] != null || latestExerciseTimeNode.Attributes["id"] != null) 
                {
                    if (latestExerciseTimeNode.Attributes["id"] != null) 
                    {
                        latestExerciseTimeIDRef_ = latestExerciseTimeNode.Attributes["id"].Value;
                        PrevailingTime ob = new PrevailingTime(latestExerciseTimeNode);
                        IDManager.SetID(latestExerciseTimeIDRef_, ob);
                    }
                    else if (latestExerciseTimeNode.Attributes["href"] != null)
                    {
                        latestExerciseTimeIDRef_ = latestExerciseTimeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        latestExerciseTime_ = new PrevailingTime(latestExerciseTimeNode);
                    }
                }
                else
                {
                    latestExerciseTime_ = new PrevailingTime(latestExerciseTimeNode);
                }
            }
            
        
            XmlNode expirationTimeNode = xmlNode.SelectSingleNode("expirationTime");
            
            if (expirationTimeNode != null)
            {
                if (expirationTimeNode.Attributes["href"] != null || expirationTimeNode.Attributes["id"] != null) 
                {
                    if (expirationTimeNode.Attributes["id"] != null) 
                    {
                        expirationTimeIDRef_ = expirationTimeNode.Attributes["id"].Value;
                        PrevailingTime ob = new PrevailingTime(expirationTimeNode);
                        IDManager.SetID(expirationTimeIDRef_, ob);
                    }
                    else if (expirationTimeNode.Attributes["href"] != null)
                    {
                        expirationTimeIDRef_ = expirationTimeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expirationTime_ = new PrevailingTime(expirationTimeNode);
                    }
                }
                else
                {
                    expirationTime_ = new PrevailingTime(expirationTimeNode);
                }
            }
            
        
        }
        
    
        #region CommencementDates_
        private AdjustableOrRelativeDates commencementDates_;
        public AdjustableOrRelativeDates CommencementDates_
        {
            get
            {
                if (this.commencementDates_ != null)
                {
                    return this.commencementDates_; 
                }
                else if (this.commencementDatesIDRef_ != null)
                {
                    commencementDates_ = IDManager.getID(commencementDatesIDRef_) as AdjustableOrRelativeDates;
                    return this.commencementDates_; 
                }
                else
                {
                      return this.commencementDates_; 
                }
            }
            set
            {
                if (this.commencementDates_ != value)
                {
                    this.commencementDates_ = value;
                }
            }
        }
        #endregion
        
        public string commencementDatesIDRef_ { get; set; }
        #region ExpirationDates_
        private AdjustableOrRelativeDates expirationDates_;
        public AdjustableOrRelativeDates ExpirationDates_
        {
            get
            {
                if (this.expirationDates_ != null)
                {
                    return this.expirationDates_; 
                }
                else if (this.expirationDatesIDRef_ != null)
                {
                    expirationDates_ = IDManager.getID(expirationDatesIDRef_) as AdjustableOrRelativeDates;
                    return this.expirationDates_; 
                }
                else
                {
                      return this.expirationDates_; 
                }
            }
            set
            {
                if (this.expirationDates_ != value)
                {
                    this.expirationDates_ = value;
                }
            }
        }
        #endregion
        
        public string expirationDatesIDRef_ { get; set; }
        #region RelativeCommencementDates_
        private CommodityRelativeExpirationDates relativeCommencementDates_;
        public CommodityRelativeExpirationDates RelativeCommencementDates_
        {
            get
            {
                if (this.relativeCommencementDates_ != null)
                {
                    return this.relativeCommencementDates_; 
                }
                else if (this.relativeCommencementDatesIDRef_ != null)
                {
                    relativeCommencementDates_ = IDManager.getID(relativeCommencementDatesIDRef_) as CommodityRelativeExpirationDates;
                    return this.relativeCommencementDates_; 
                }
                else
                {
                      return this.relativeCommencementDates_; 
                }
            }
            set
            {
                if (this.relativeCommencementDates_ != value)
                {
                    this.relativeCommencementDates_ = value;
                }
            }
        }
        #endregion
        
        public string relativeCommencementDatesIDRef_ { get; set; }
        #region RelativeExpirationDates_
        private CommodityRelativeExpirationDates relativeExpirationDates_;
        public CommodityRelativeExpirationDates RelativeExpirationDates_
        {
            get
            {
                if (this.relativeExpirationDates_ != null)
                {
                    return this.relativeExpirationDates_; 
                }
                else if (this.relativeExpirationDatesIDRef_ != null)
                {
                    relativeExpirationDates_ = IDManager.getID(relativeExpirationDatesIDRef_) as CommodityRelativeExpirationDates;
                    return this.relativeExpirationDates_; 
                }
                else
                {
                      return this.relativeExpirationDates_; 
                }
            }
            set
            {
                if (this.relativeExpirationDates_ != value)
                {
                    this.relativeExpirationDates_ = value;
                }
            }
        }
        #endregion
        
        public string relativeExpirationDatesIDRef_ { get; set; }
        #region LatestExerciseTime_
        private PrevailingTime latestExerciseTime_;
        public PrevailingTime LatestExerciseTime_
        {
            get
            {
                if (this.latestExerciseTime_ != null)
                {
                    return this.latestExerciseTime_; 
                }
                else if (this.latestExerciseTimeIDRef_ != null)
                {
                    latestExerciseTime_ = IDManager.getID(latestExerciseTimeIDRef_) as PrevailingTime;
                    return this.latestExerciseTime_; 
                }
                else
                {
                      return this.latestExerciseTime_; 
                }
            }
            set
            {
                if (this.latestExerciseTime_ != value)
                {
                    this.latestExerciseTime_ = value;
                }
            }
        }
        #endregion
        
        public string latestExerciseTimeIDRef_ { get; set; }
        #region ExpirationTime_
        private PrevailingTime expirationTime_;
        public PrevailingTime ExpirationTime_
        {
            get
            {
                if (this.expirationTime_ != null)
                {
                    return this.expirationTime_; 
                }
                else if (this.expirationTimeIDRef_ != null)
                {
                    expirationTime_ = IDManager.getID(expirationTimeIDRef_) as PrevailingTime;
                    return this.expirationTime_; 
                }
                else
                {
                      return this.expirationTime_; 
                }
            }
            set
            {
                if (this.expirationTime_ != value)
                {
                    this.expirationTime_ = value;
                }
            }
        }
        #endregion
        
        public string expirationTimeIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/CommodityPhysicalAmericanExercise.cpp
// CommodityPhysicalAmericanExercise.cpp 
#include "CommodityPhysicalAmericanExercise.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CommodityPhysicalAmericanExercise::CommodityPhysicalAmericanExercise(TiXmlNode* xmlNode)
: Exercise(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //commencementDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* commencementDatesNode = xmlNode->FirstChildElement("commencementDates");

   if(commencementDatesNode){commencementDatesIsNull_ = false;}
   else{commencementDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- commencementDatesNode , address : " << commencementDatesNode << std::endl;
   #endif
   if(commencementDatesNode)
   {
      if(commencementDatesNode->Attribute("href") || commencementDatesNode->Attribute("id"))
      {
          if(commencementDatesNode->Attribute("id"))
          {
             commencementDatesIDRef_ = commencementDatesNode->Attribute("id");
             commencementDates_ = boost::shared_ptr<AdjustableOrRelativeDates>(new AdjustableOrRelativeDates(commencementDatesNode));
             commencementDates_->setID(commencementDatesIDRef_);
             IDManager::instance().setID(commencementDatesIDRef_,commencementDates_);
          }
          else if(commencementDatesNode->Attribute("href")) { commencementDatesIDRef_ = commencementDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { commencementDates_ = boost::shared_ptr<AdjustableOrRelativeDates>(new AdjustableOrRelativeDates(commencementDatesNode));}
   }

   //expirationDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expirationDatesNode = xmlNode->FirstChildElement("expirationDates");

   if(expirationDatesNode){expirationDatesIsNull_ = false;}
   else{expirationDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expirationDatesNode , address : " << expirationDatesNode << std::endl;
   #endif
   if(expirationDatesNode)
   {
      if(expirationDatesNode->Attribute("href") || expirationDatesNode->Attribute("id"))
      {
          if(expirationDatesNode->Attribute("id"))
          {
             expirationDatesIDRef_ = expirationDatesNode->Attribute("id");
             expirationDates_ = boost::shared_ptr<AdjustableOrRelativeDates>(new AdjustableOrRelativeDates(expirationDatesNode));
             expirationDates_->setID(expirationDatesIDRef_);
             IDManager::instance().setID(expirationDatesIDRef_,expirationDates_);
          }
          else if(expirationDatesNode->Attribute("href")) { expirationDatesIDRef_ = expirationDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expirationDates_ = boost::shared_ptr<AdjustableOrRelativeDates>(new AdjustableOrRelativeDates(expirationDatesNode));}
   }

   //relativeCommencementDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* relativeCommencementDatesNode = xmlNode->FirstChildElement("relativeCommencementDates");

   if(relativeCommencementDatesNode){relativeCommencementDatesIsNull_ = false;}
   else{relativeCommencementDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- relativeCommencementDatesNode , address : " << relativeCommencementDatesNode << std::endl;
   #endif
   if(relativeCommencementDatesNode)
   {
      if(relativeCommencementDatesNode->Attribute("href") || relativeCommencementDatesNode->Attribute("id"))
      {
          if(relativeCommencementDatesNode->Attribute("id"))
          {
             relativeCommencementDatesIDRef_ = relativeCommencementDatesNode->Attribute("id");
             relativeCommencementDates_ = boost::shared_ptr<CommodityRelativeExpirationDates>(new CommodityRelativeExpirationDates(relativeCommencementDatesNode));
             relativeCommencementDates_->setID(relativeCommencementDatesIDRef_);
             IDManager::instance().setID(relativeCommencementDatesIDRef_,relativeCommencementDates_);
          }
          else if(relativeCommencementDatesNode->Attribute("href")) { relativeCommencementDatesIDRef_ = relativeCommencementDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { relativeCommencementDates_ = boost::shared_ptr<CommodityRelativeExpirationDates>(new CommodityRelativeExpirationDates(relativeCommencementDatesNode));}
   }

   //relativeExpirationDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* relativeExpirationDatesNode = xmlNode->FirstChildElement("relativeExpirationDates");

   if(relativeExpirationDatesNode){relativeExpirationDatesIsNull_ = false;}
   else{relativeExpirationDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- relativeExpirationDatesNode , address : " << relativeExpirationDatesNode << std::endl;
   #endif
   if(relativeExpirationDatesNode)
   {
      if(relativeExpirationDatesNode->Attribute("href") || relativeExpirationDatesNode->Attribute("id"))
      {
          if(relativeExpirationDatesNode->Attribute("id"))
          {
             relativeExpirationDatesIDRef_ = relativeExpirationDatesNode->Attribute("id");
             relativeExpirationDates_ = boost::shared_ptr<CommodityRelativeExpirationDates>(new CommodityRelativeExpirationDates(relativeExpirationDatesNode));
             relativeExpirationDates_->setID(relativeExpirationDatesIDRef_);
             IDManager::instance().setID(relativeExpirationDatesIDRef_,relativeExpirationDates_);
          }
          else if(relativeExpirationDatesNode->Attribute("href")) { relativeExpirationDatesIDRef_ = relativeExpirationDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { relativeExpirationDates_ = boost::shared_ptr<CommodityRelativeExpirationDates>(new CommodityRelativeExpirationDates(relativeExpirationDatesNode));}
   }

   //latestExerciseTimeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* latestExerciseTimeNode = xmlNode->FirstChildElement("latestExerciseTime");

   if(latestExerciseTimeNode){latestExerciseTimeIsNull_ = false;}
   else{latestExerciseTimeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- latestExerciseTimeNode , address : " << latestExerciseTimeNode << std::endl;
   #endif
   if(latestExerciseTimeNode)
   {
      if(latestExerciseTimeNode->Attribute("href") || latestExerciseTimeNode->Attribute("id"))
      {
          if(latestExerciseTimeNode->Attribute("id"))
          {
             latestExerciseTimeIDRef_ = latestExerciseTimeNode->Attribute("id");
             latestExerciseTime_ = boost::shared_ptr<PrevailingTime>(new PrevailingTime(latestExerciseTimeNode));
             latestExerciseTime_->setID(latestExerciseTimeIDRef_);
             IDManager::instance().setID(latestExerciseTimeIDRef_,latestExerciseTime_);
          }
          else if(latestExerciseTimeNode->Attribute("href")) { latestExerciseTimeIDRef_ = latestExerciseTimeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { latestExerciseTime_ = boost::shared_ptr<PrevailingTime>(new PrevailingTime(latestExerciseTimeNode));}
   }

   //expirationTimeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expirationTimeNode = xmlNode->FirstChildElement("expirationTime");

   if(expirationTimeNode){expirationTimeIsNull_ = false;}
   else{expirationTimeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expirationTimeNode , address : " << expirationTimeNode << std::endl;
   #endif
   if(expirationTimeNode)
   {
      if(expirationTimeNode->Attribute("href") || expirationTimeNode->Attribute("id"))
      {
          if(expirationTimeNode->Attribute("id"))
          {
             expirationTimeIDRef_ = expirationTimeNode->Attribute("id");
             expirationTime_ = boost::shared_ptr<PrevailingTime>(new PrevailingTime(expirationTimeNode));
             expirationTime_->setID(expirationTimeIDRef_);
             IDManager::instance().setID(expirationTimeIDRef_,expirationTime_);
          }
          else if(expirationTimeNode->Attribute("href")) { expirationTimeIDRef_ = expirationTimeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expirationTime_ = boost::shared_ptr<PrevailingTime>(new PrevailingTime(expirationTimeNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<AdjustableOrRelativeDates> CommodityPhysicalAmericanExercise::getCommencementDates()
{
   if(!this->commencementDatesIsNull_){
        if(commencementDatesIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDates>(IDManager::instance().getID(commencementDatesIDRef_));
        }else{
             return this->commencementDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDates>();
   }
}
boost::shared_ptr<AdjustableOrRelativeDates> CommodityPhysicalAmericanExercise::getExpirationDates()
{
   if(!this->expirationDatesIsNull_){
        if(expirationDatesIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDates>(IDManager::instance().getID(expirationDatesIDRef_));
        }else{
             return this->expirationDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDates>();
   }
}
boost::shared_ptr<CommodityRelativeExpirationDates> CommodityPhysicalAmericanExercise::getRelativeCommencementDates()
{
   if(!this->relativeCommencementDatesIsNull_){
        if(relativeCommencementDatesIDRef_ != ""){
             return boost::shared_static_cast<CommodityRelativeExpirationDates>(IDManager::instance().getID(relativeCommencementDatesIDRef_));
        }else{
             return this->relativeCommencementDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CommodityRelativeExpirationDates>();
   }
}
boost::shared_ptr<CommodityRelativeExpirationDates> CommodityPhysicalAmericanExercise::getRelativeExpirationDates()
{
   if(!this->relativeExpirationDatesIsNull_){
        if(relativeExpirationDatesIDRef_ != ""){
             return boost::shared_static_cast<CommodityRelativeExpirationDates>(IDManager::instance().getID(relativeExpirationDatesIDRef_));
        }else{
             return this->relativeExpirationDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CommodityRelativeExpirationDates>();
   }
}
boost::shared_ptr<PrevailingTime> CommodityPhysicalAmericanExercise::getLatestExerciseTime()
{
   if(!this->latestExerciseTimeIsNull_){
        if(latestExerciseTimeIDRef_ != ""){
             return boost::shared_static_cast<PrevailingTime>(IDManager::instance().getID(latestExerciseTimeIDRef_));
        }else{
             return this->latestExerciseTime_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PrevailingTime>();
   }
}
boost::shared_ptr<PrevailingTime> CommodityPhysicalAmericanExercise::getExpirationTime()
{
   if(!this->expirationTimeIsNull_){
        if(expirationTimeIDRef_ != ""){
             return boost::shared_static_cast<PrevailingTime>(IDManager::instance().getID(expirationTimeIDRef_));
        }else{
             return this->expirationTime_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PrevailingTime>();
   }
}
}

